common/headers/headers.go

```diff
@@ -0,0 +1,100 @@
+package headers
+
+import (
+	"context"
+
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	// LibraryVersionHeaderName refers to the name of the
+	// tchannel / http header that contains the client
+	// library version
+	LibraryVersionHeaderName = "cadence-client-library-version"
+
+	// FeatureVersionHeaderName refers to the name of the
+	// tchannel / http header that contains the client
+	// feature version
+	// the feature version sent from client represents the
+	// feature set of the cadence client library support.
+	// This can be used for client capibility check, on
+	// Cadence server, for backward compatibility
+	FeatureVersionHeaderName = "cadence-client-feature-version"
+
+	// ClientImplHeaderName refers to the name of the
+	// header that contains the client implementation
+	ClientImplHeaderName = "cadence-client-name"
+	// EnforceDCRedirection refers to a boolean string of whether
+	// to enforce DCRedirection(auto-forwarding)
+	// Will be removed in the future: https://github.com/uber/cadence/issues/2304
+	EnforceDCRedirection = "cadence-enforce-dc-redirection"
+)
+
+// GetValues returns header values for passed header names.
+// It always returns slice of the same size as number of passed header names.
+func GetValues(ctx context.Context, headerNames ...string) []string {
+	headerValues := make([]string, len(headerNames))
+
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		for i, headerName := range headerNames {
+			headerValues[i] = getSingleHeaderValue(md, headerName)
+		}
+	}
+
+	return headerValues
+}
+
+// PropagateVersions propagates headers from incoming context to outgoing context.
+func PropagateVersions(ctx context.Context) context.Context {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		outgoingMetadata := copyIncomingHeadersToOutgoing(md,
+			LibraryVersionHeaderName,
+			FeatureVersionHeaderName,
+			ClientImplHeaderName)
+		if outgoingMetadata.Len() > 0 {
+			ctx = metadata.NewOutgoingContext(ctx, outgoingMetadata)
+		}
+	}
+
+	return ctx
+}
+
+// SetVersions sets headers for internal communications.
+func SetVersions(ctx context.Context) context.Context {
```

Comment:
Why SetVersions and not SetHeaders?

Reply:
Because package name is `headers`. So on caller side it looks like `headers.SetVersions`.

Reply:
Are these always going to be versions? I see we currently have `EnforceDCRedirection` (though it will be deprecated eventually) Should this be .SetProperties or SetFlags to keep generic verbiage?

Reply:
This func sets version headers only. `EnforceDCRedirection` is set separately.